[DBZ-3506] Remove SchemaProcessor from Cassandra Connector

bingqinzhou opened the pull request:
https://issues.redhat.com/browse/DBZ-3506

- Remove SchemaProcessor from Cassandra Connector 
- Clean up after PR#26
- Update Doc in https://github.com/debezium/debezium/pull/2396

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,292 @@
+            }
+        }
+
+        @Override
+        public void onUserTypeAdded(final UserType type) {
+        }
```
Why all the empty implementations?

bingqinzhou replied:
> Why all the empty implementations?

From my understanding, this is a required method to be overriden to implement SchemaChangeListener in Cassandra, but Cassandra Connector doesn't actually need to do anything for the case onUserTypeAdded so the function is listed here but kept empty.

gunnarmorling replied:
Yes, makes sense. Thx! I'd suggest to put comments in there:

    // not relevant to Debezium

bingqinzhou replied:
> Yes, makes sense. Thx! I'd suggest to put comments in there:
> 
> ```
> // not relevant to Debezium
> ```

Added comments accordingly.

gunnarmorling reviewed (commented):
One question inline. Got to admit I'm still not up-to-speed with all the new support for schema changes yet :(

gunnarmorling commented:
@bingqinzhou, could you rebase this one please to the latest master? In general, we prefer rebases over merges for the sake of a linear (and thus easier to follow) history. Thx!

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
 /**
- * Caches the key and value schema for all CDC-enabled tables. This cache gets updated
- * by {@link SchemaProcessor} periodically.
+ * Listening to schema changes in Cassandra DB and caches the key and value schema for all CDC-enabled tables.
+ * This cache gets updated whenever there's a schema change in Cassandra DB
```
A general question on the handling of schema changes; IIUC, the connector is triggered via a listener in case there's a schema change; how is this sync-ed with the log reading process? E.g. when reading changes from a log file created _before_ that schema change, how are we making sure that those parsed events are propagated with the previous, then valid schema?

bingqinzhou replied:
This is not related to this PR, but related to PR#26. I left a comment in DBZ-3417 to answer this question and explain about the trade off: https://issues.redhat.com/browse/DBZ-3417?focusedCommentId=16227506&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16227506

bingqinzhou replied:
> A general question on the handling of schema changes; IIUC, the connector is triggered via a listener in case there's a schema change; how is this sync-ed with the log reading process? E.g. when reading changes from a log file created _before_ that schema change, how are we making sure that those parsed events are propagated with the previous, then valid schema?

This is not related to this PR, but related to PR#26. I left a comment in DBZ-3417 to answer this question and explain about the trade off in detail: https://issues.redhat.com/browse/DBZ-3417?focusedCommentId=16227506&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16227506

gunnarmorling replied:
Ok, thx! Also thx a lot for explaining this to me in the call yesterday!

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
-    public final SourceInfoStructMaker sourceInfoStructMaker;
+    private final Map<KeyspaceTable, KeyValueSchema> tableToKVSchemaMap;
+    private final String kafkaTopicPrefix;
+    private final SourceInfoStructMaker sourceInfoStructMaker;
```
```suggestion
    private final SourceInfoStructMaker<SourceInfo> sourceInfoStructMaker;
```

bingqinzhou replied:
> 

Updated.

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
 
-    public final String kafkaTopicPrefix;
-    public final SourceInfoStructMaker sourceInfoStructMaker;
+    private final Map<KeyspaceTable, KeyValueSchema> tableToKVSchemaMap;
```
```suggestion
    private final ConcurrentMap<KeyspaceTable, KeyValueSchema> tableToKVSchemaMap;
```

bingqinzhou replied:
> 

Updated.

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                        KeyspaceParams.create(keyspaceMetadata.isDurableWrites(),
+                                keyspaceMetadata.getReplication())));
+                Keyspace.openWithoutSSTables(keyspaceMetadata.getName());
+                LOGGER.info("Added keyspace [{}] to Schema Instance.", keyspaceMetadata.asCQLQuery());
```
```suggestion
                LOGGER.info("Added keyspace [{}] to schema instance.", keyspaceMetadata.asCQLQuery());
```

bingqinzhou replied:
> 

Updated this and the following logs.

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                LOGGER.info("Added keyspace [{}] to Schema Instance.", keyspaceMetadata.asCQLQuery());
+            }
+            catch (Exception e) {
+                LOGGER.warn("Error happened while adding the keyspace {} to Schema instance", keyspaceMetadata.getName(), e);
```
```suggestion
                LOGGER.warn("Error happened while adding the keyspace {} to schema instance", keyspaceMetadata.getName(), e);
```

So should we truly swallow this? Or rather bubble up? I.e. should the connector actually keep going? What will happen later on when parsing a change event for that key space?

bingqinzhou replied:
Again this is not related to this PR, I was just copying the codes from SchemaProcessor in PR#26. If Cassandra Connector stops, it will potentially cause write suspension in Cassandra DB. For this reason, we want to avoid throwing exceptions in this connector as much as possible.

bingqinzhou replied:
> So should we truly swallow this? Or rather bubble up? I.e. should the connector actually keep going? What will happen later on when parsing a change event for that key space?

Again this is not related to this PR, I was just copying the codes from SchemaProcessor in PR#26. If Cassandra Connector stops, it will potentially cause write suspension in Cassandra DB. For this reason, we want to avoid throwing exceptions in this connector as much as possible.

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                LOGGER.info("Updated keyspace [{}] in Schema Instance.", current.asCQLQuery());
+            }
+            catch (Exception e) {
+                LOGGER.warn("Error happened while updating the keyspace {} in Schema instance", current.getName(), e);
```
```suggestion
                LOGGER.warn("Error happened while updating the keyspace {} in schema instance", current.getName(), e);
```

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                LOGGER.info("Removed keyspace [{}] from Schema Instance.", keyspaceMetadata.asCQLQuery());
+            }
+            catch (Exception e) {
+                LOGGER.warn("Error happened while removing the keyspace {} from Schema instance", keyspaceMetadata.getName(), e);
```
```suggestion
                LOGGER.warn("Error happened while removing the keyspace {} from schema instance", keyspaceMetadata.getName(), e);
```

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                    LOGGER.debug("Table {}.{} is already added!", tableMetadata.getKeyspace(), tableMetadata.getName());
+                    return;
+                }
+                final java.util.function.Function<org.apache.cassandra.schema.KeyspaceMetadata, org.apache.cassandra.schema.KeyspaceMetadata> transformationFunction = ks -> ks
+                        .withSwapped(ks.tables.with(newCFMetaData));
+                org.apache.cassandra.schema.KeyspaceMetadata transformed = transformationFunction.apply(current);
```
Why going through the `Function` indirection here?

bingqinzhou replied:
> Why going through the `Function` indirection here?

I'm not sure about this, just copying the codes from SchemaProcessor landed in PR#26.

gunnarmorling replied:
Ok, will push a commit for simplifying this and merge. Thx!

gunnarmorling commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -5,154 +5,305 @@
+                org.apache.cassandra.schema.KeyspaceMetadata transformed = transformationFunction.apply(current);
+                Schema.instance.setKeyspaceMetadata(transformed);
+                Schema.instance.load(newCFMetaData);
+                LOGGER.info("Added table [{}] to Schema Instance.", tableMetadata.asCQLQuery());
```
Same as above (and below). Let's try and make the casing consistent.

gunnarmorling reviewed (commented):
@bingqinzhou, some more comments after taking another look.

bingqinzhou commented:
> @bingqinzhou, could you rebase this one please to the latest master? In general, we prefer rebases over merges for the sake of a linear (and thus easier to follow) history. Thx!

Rebased and squashed commits.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/KeyValueSchema.java:
```diff
@@ -0,0 +1,93 @@
+import io.debezium.connector.cassandra.transforms.CassandraTypeConverter;
+import io.debezium.connector.cassandra.transforms.CassandraTypeDeserializer;
+
+public class KeyValueSchema {
```
> Please add a comment describing the classes purpose. Let's try and make this a habit in general.

Added comments accordingly.

gunnarmorling commented:
Reworked a bit to get rid of the merge conflicts and added one more commit for simplifying that `Function` usage mentioned above. Thx, @bingqinzhou!
